### PR TITLE
Add pixel depth rounding to d3d9, gles2 and support forcing via config

### PR DESCRIPTION
**Pull request by unknownbrackets**
This will also work for GLES2 devices.  Since depth is fixed point, dividing the range should be mostly accurate - only problem is the floats sent to the depth range func, really.

Code changes overlap with #8447, but I've intentionally kept it so both changes work together.  Conflict is easy to resolve - just keep both changes.

Helps #6485 when Direct3D9 or GLES2 is used.

-[Unknown]

### CODE REVIEW

File: GPU/GLES/GLES_GPU.cpp [hunk 1]
```diff
@@ -565,11 +565,21 @@ void GLES_GPU::CheckGPUFeatures() {
...
-	} else {
-		if (PSP_CoreParameter().compat.flags().VertexDepthRounding) {
+	// If we already have a 16-bit depth buffer, we don't need to round.
+	if (fbo_standard_z_depth() > 16) {
+		if (PSP_CoreParameter().compat.flags().PixelDepthRounding) {
+			if (!gl_extensions.IsGLES || gl_extensions.GLES3) {
+				// Use fragment rounding on desktop and GLES3, most accurate.
+				features |= GPU_ROUND_FRAGMENT_DEPTH_TO_16BIT;
+			} else if (fbo_standard_z_depth() == 24) {
+				// Here we can simulate a 16 bit depth buffer by clamping.
```
Comment (hrydgard) [inline GPU/GLES/GLES_GPU.cpp]: I'm not sure I would call this clamping - rather just mapping to a reduce range. although CLAMP is shorter than anything I can come up with for the latter though :)

**Comment by unknownbrackets** [comment]
By the way (maybe because of precision) but `GPU_CLAMP_DEPTH_FROM_24BIT_TO_16BIT` chases the evil MotoGP cloud away.  `GPU_ROUND_FRAGMENT_DEPTH_TO_16BIT` does not, and neither does `GPU_ROUND_DEPTH_TO_16BIT`.

-[Unknown]

**Comment by hrydgard** [comment]
Could be that your range calculation for GPU_CLAMP_DEPTH_... accidentally implements that 0.5 offset?

File: GPU/GLES/GLES_GPU.cpp (hunk 1 above)
Comment (unknownbrackets) [inline GPU/GLES/GLES_GPU.cpp]: Ah, you're right.  I've changed it to SCALE.

-[Unknown]

**Comment by unknownbrackets** [comment]
Hmm.  Actually, right.  That is what's happening.

The negative values now map to usable depth values between `0.0` and `1.0` in GL-land, so it doesn't even bother clamping at all.  The `zOffset` and `zScale` don't get set.  So in other words, this makes it not clip properly - I still need to clamp the near/far to those values...

I should rebase this on #8447, because the code'll be simpler.

-[Unknown]

**Comment by unknownbrackets** [comment]
Rebased.  Back to MotoGP not working as expected.

-[Unknown]

**Comment by unknownbrackets** [comment]
I've put a bunch of things in here, I'll rebase it once some other depth changes happen, but I'd still like to have an option for the reduced depth range, as in these changes.

-[Unknown]

**Comment by unknownbrackets** [comment]
While testing this, I noticed that per-vertex rounding in Phantasy Star Portable 1 isn't showing the text.  It works fine with per-pixel rounding (this feature.)  Maybe we should just kill per-vertex rounding in d3d9?

After debugging it, I've decided that though and non through are rounding differently.  It's an offset of 0.015625, which is 256/4, which means it's off by 1 (at fixed point 24.)  It works perfectly in both vertex and pixel rounding if I add a half pixel offset.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah maybe we can even kill per-vertex rounding completely now that we have this?

**Comment by unknownbrackets** [comment]
Yeah.  I mean, per-vertex rounding might allow you to retain more depth quality, but it's also more complicated.  We could avoid the u_depthRange update entirely if we removed it - saving some vertex uniforms.

I guess the only thing is that per-vertex rounding is faster than per-pixel, if you have GLES3 support.

-[Unknown]

**Comment by hrydgard** [comment]
Actually, is there still any use for the pixel shader depth rounding - can't limited depth range simulate is pretty much completely?

**Comment by unknownbrackets** [comment]
Yes, assuming drivers don't have weird precision issues.  It also has the added benefit of having an increased "clamp" range.

We could still potentially clamp in the fragment shader, in the same place we're rounding now, if some game depends on it.  Seems unlikely...

But yeah, I think it makes sense to just remove the other options.  Also simplifies the compat options.

One concern I did think of - devices with native 16-bit buffers.  Currently, we will cut those down to 14-bit, although we won't go lower (pixel/vertex rounding will simply not enable.)  Maybe clamping is the better thing to lose...

-[Unknown]

**Comment by daniel229** [comment]
Taiko no Tatsujin Portable DX doesn't work fine with PixelDepthRounding.

**Pull request merged**